src/main/java/com/cloudbees/hudson/plugins/folder/icons/BuildStatusFolderIcon.java

```diff
@@ -0,0 +1,84 @@
+package com.cloudbees.hudson.plugins.folder.icons;
+
+import com.cloudbees.hudson.plugins.folder.Folder;
+import com.cloudbees.hudson.plugins.folder.FolderIcon;
+import com.cloudbees.hudson.plugins.folder.FolderIconDescriptor;
+import hudson.Extension;
+import hudson.model.*;
+import org.kohsuke.stapler.DataBoundConstructor;
+import org.kohsuke.stapler.Stapler;
+
+/**
+ * Created by dune on 09/09/15.
+ */
+
+    public class BuildStatusFolderIcon extends FolderIcon {
+
+        Folder folder;
+
+        @DataBoundConstructor
+        public BuildStatusFolderIcon() {
+        }
+
+        public Integer ordinal() {
+          return this.getCombinedBallColor().ordinal();
+        }
+
+        private BallColor getCombinedBallColor() {
+            Boolean taskIsBuilding = false;
+            Boolean oneTaskIsIncomplete = false;
+            Result combinedResult = null;
+            for(Job job : this.folder.getAllJobs()) {
+                Run build = job.getLastBuild();
+                if(build == null) {
+                    oneTaskIsIncomplete = true;
+                    continue;
+                }
+
+                if(build.isBuilding()) {
+                    taskIsBuilding = true;
+                    build = job.getLastCompletedBuild();
+                    if(build == null) continue;
```

Comment:
Based on how the icon gets calculated, adding the following here will speed up the calculation:

```
  if (oneTaskIsIncomplete) {
    break;
  }
```